runtime/src/main/cpp/Memory.cpp

```diff
@@ -1604,38 +1629,77 @@ void garbageCollect(MemoryState* state, bool force) {
   if (g_hasCyclicCollector)
     cyclicLocalGC();
 #endif  // USE_CYCLIC_GC
+#if PROFILE_GC
+  auto processDecrementsStartTime = konan::getTimeMicros();
+#endif
   processDecrements(state);
+#if PROFILE_GC
+  auto processDecrementsDuration = konan::getTimeMicros() - processDecrementsStartTime;
+  GC_LOG("||| GC: processDecrementsDuration = %lld\n", processDecrementsDuration);
+  auto decrementStackStartTime = konan::getTimeMicros();
+#endif
   size_t beforeDecrements = state->toRelease->size();
   decrementStack(state);
   size_t afterDecrements = state->toRelease->size();
+#if PROFILE_GC
+  auto decrementStackDuration = konan::getTimeMicros() - decrementStackStartTime;
+  GC_LOG("||| GC: decrementStackDuration = %lld\n", decrementStackDuration);
+#endif
   long stackReferences = afterDecrements - beforeDecrements;
   if (state->gcErgonomics && stackReferences * 5 > state->gcThreshold) {
     increaseGcThreshold(state);
-    GC_LOG("||| GC: too many stack references, increased threshold to \n", state->gcThreshold);
+    GC_LOG("||| GC: too many stack references, increased threshold to %d\n", state->gcThreshold);
   }
 
   GC_LOG("||| GC: toFree %d toRelease %d\n", state->toFree->size(), state->toRelease->size())
-
+#if PROFILE_GC
+  auto processFinalizerQueueStartTime = konan::getTimeMicros();
+#endif
   processFinalizerQueue(state);
+#if PROFILE_GC
+  auto processFinalizerQueueDuration = konan::getTimeMicros() - processFinalizerQueueStartTime;
+  GC_LOG("||| GC: processFinalizerQueueDuration %lld\n", processFinalizerQueueDuration);
+#endif
 
-  if (force || state->toFree->size() > kMaxToFreeSize) {
+  int64_t collectCyclesDuration = 0;
+  int64_t cyclicGcEndTime = 0;
+  if (force || state->toFree->size() > state->gcCollectCyclesThreshold) {
     while (state->toFree->size() > 0) {
+      auto collectCyclesStartTime = konan::getTimeMicros();
```

Comment:
I'm wondering: what if we include `processFinalizerQueue` execution time from below into `collectCycles` time? This way we can wrap the loop in `cyclicGcStartTime = ...; while(...) {...}; cyclicGcEndTime = ...;` and adjust the `if` below `double(cyclicGcEndTime - cyclicGcStartTime) / (cyclicGcStartTime - state->lastCyclicGcTimestamp + 1)`. Rationale behind this: `processFinalizerQueue` work is done precisely because we were able to resolve some cycles (or am I wrong here?). What do you think?

Reply:
I think it would be fine for current approach, if there is a lot of garbage then  just freeing takes quite a big amount of time. But usually there is small amount of cyclic garbage. I tried this to use less calls for measuring time and check on benchmarks.

Reply:
IMHO, measuring only `collectCycles` seems more precise, because processing finalizer queue is always a "useful work". If cycle collection takes 1 ms while processing finalizer queue takes 1000 ms, the latter shouldn't be the reason to perform the former less often.
On the other hand, T(cycle collection) > kT(process finalizer queue). So this example is not exactly possible.

Considering the fact that measuring processing finalizer queue makes the code simpler, I guess it is ok to keep it that way.

Reply:
I agree that it's more precise, but impact of processing finalizer queue is quite small. Nevertheless I checked on benchmarks, nothing critical because of this change.